Add vector directive

mjfwebb opened the pull request:
# Description

This PR introduces a new directive: `@vector`. This feature requires a Neo4j database version of at least 5.15.

The `@vector` directive, with an accompanying [vector index](https://neo4j.com/docs/cypher-manual/current/indexes/semantic-indexes/vector-indexes/) enables searching for nodes with a similar vector.

The directive enables two forms of user input, depending on the index configuration: `vector` input which in GraphQL is a `[Float!]`, and `phrase` input which is a `String`.

Phrase input utilises the [Neo4j GenAI plugin](https://neo4j.com/docs/cypher-manual/current/genai-integrations/) functionality available on the Neo4j database. This feature is enabled by default in [Neo4j Aura](https://neo4j.com/cloud/aura-free/).

## Vector input

With vector input, the API user provides a list of floats (`[Float!]`) which are used as input to Cypher [query nodes](https://neo4j.com/docs/cypher-manual/current/indexes/semantic-indexes/vector-indexes/#query-vector-index) against an existing vector index. 

Consider the following type definition:
```graphql
type Movie
    @vector(
        indexes: [
            { indexName: "myVectorIndexName", propertyName: "embedding", queryName: "searchForRelatedMovies" }
        ]
    ) {
    title: String!
    released: Int!
    actors: [Person!]! @relationship(type: "ACTED_IN", direction: IN)
}
```

The addition of the `@vector` directive means you users can perform a query with a vector as input:

Query:
```graphql
query($inputVector: [Float!]!) {
  searchForRelatedMovies(vector: $inputVector) {
    moviesConnection {
      edges {
        node {
          title
        }
        score
      }
      totalCount
    }
  }
}
```
Input (truncated):
```json
{
  "inputVector": [
    0.123456,
    ...,
    0.654321,
  ]
}
```

## Phrase Input

With phrase input, the API user provides a phrase (`String!`) which is used as input to utilise the [Neo4j GenAI plugin](https://neo4j.com/docs/cypher-manual/current/genai-integrations/), before [querying nodes](https://neo4j.com/docs/cypher-manual/current/indexes/semantic-indexes/vector-indexes/#query-vector-index) against an existing vector index. 

Usage of the GenAI plugin is enabled by adding a provider (enum) to the index configuration in the type definitions, and having an associated configuration supplied to the Neo4jGraphQL constructor under the `vector` feature.

Consider the following type definition:
```graphql
type Movie
    @vector(
        indexes: [
            { indexName: "myVectorIndexName", propertyName: "embedding", queryName: "searchForRelatedMovies", provider: OPEN_AI }
        ]
    ) {
    title: String!
    released: Int!
    actors: [Person!]! @relationship(type: "ACTED_IN", direction: IN)
}
```
Combined with the following Neo4jGraphQL constructor configuration where the configuration matches the enum set as the provider in the index (OPEN_AI to OpenAI):
```ts
const neoSchema = new Neo4jGraphQL({
    typeDefs,
    driver,
    features: {
        vector: {
            OpenAI: {
                token: "my-open-ai-token",
                model: "text-embedding-3-small",
            },
        },
    },
});
```

The addition of the `provider` means you users can perform a query with a phrase as input:
```graphql
query($inputPhrase: String!) {
  searchForRelatedMovies(phrase: $inputPhrase) {
    moviesConnection {
      edges {
        node {
          title
        }
        score
      }
      totalCount
    }
  }
}
```
Input:
```json
{
  "inputPhrase": "Friends become enemies, enemies become friends"
}
```

# Complexity

This is a new feature, including a new API (where you select edges and node).

Complexity: Medium

mjfwebb committed:
refactor: move SCORE_FIELD to constants

mjfwebb committed:
feat: add genAI directive and annotation

mjfwebb committed:
feat: add genAI input types

mjfwebb committed:
feat: add augmentGenAISchema for genAI query/input schema generation

mjfwebb committed:
feat: add vector index assertion for genai directive

mjfwebb committed:
feat: add genAI to features in Neo4jGraphQLConstructor

mjfwebb committed:
wip

mjfwebb committed:
feat: ensure vector indexes are checked when create set to true

mjfwebb committed:
wip

mjfwebb committed:
wip testing

mjfwebb committed:
wip

mjfwebb committed:
wip

mjfwebb committed:
wip

mjfwebb committed:
remove commented out code

mjfwebb committed:
wip

angrykoala committed:
Add node filters to vector operations

angrykoala committed:
Update integration tests for vector directive

angrykoala committed:
Add sorting on vector index

angrykoala committed:
Pagination with vector tests

angrykoala committed:
Move score sort to a method in vector factory

angrykoala committed:
Add test on relationship in query vector

angrykoala committed:
enables provider for genai plugin

angrykoala committed:
Merge pull request #5284 from neo4j/vector-index-test

Support for vector index filtering and sorting

mjfwebb committed:
feat: use enums for vector provider names

mjfwebb committed:
fix: update vector-annotation test

mjfwebb committed:
Merge branch 'dev' into single-vector-directive

mjfwebb committed:
test: add checks for database version in integration tests

mjfwebb committed:
test: do not verify TCK on vector tests

mjfwebb committed:
test: add vector auth integration tests

angrykoala committed:
Add changeset for edit and sorting in vector

mjfwebb committed:
Merge branch 'dev' into single-vector-directive

neo4j-team-graphql commented:
# Performance Report

No Performance Changes


<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3404 | 3404 | 36 | 43 | 1134 |
|  aggregations.TopLevelAggregateWithMultipleFields | 6802 | 6802 | 44 | 69 | 1134 |
|  aggregations.NestedAggregation | 16544 | 16544 | 74 | 95 | 2174 |
|  aggregations.AggregationWithWhere | 11970 | 11970 | 39 | 56 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 22116478 | 22116478 | 2235 | 2513 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 22116478 | 22116478 | 2256 | 2297 | 2008534 |
|  aggregations.NestedCountFromMovieToActors | 9701 | 9701 | 33 | 50 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 9928 | 9928 | 37 | 53 | 2174 |
|  aggregations.DeeplyNestedCount | 11062184 | 11062184 | 3047 | 2781 | 2008534 |
|  aggregations.InterfacesAggregations | 6242 | 6242 | 43 | 59 | 2080 |
|  aggregations.InterfacesAggregationsWithTwoFields | 11444 | 11444 | 70 | 90 | 2080 |
|  batch-create.BatchCreate | 4200 | 4200 | 121 | 141 | 600 |
|  batch-create.BatchCreateSmall | 77 | 77 | 49 | 61 | 11 |
|  connect.createAndConnect | 14424 | 14424 | 134 | 220 | 3003 |
|  connections.Connection | 14049 | 14049 | 96 | 67 | 2174 |
|  connections.NestedConnection | 41068 | 41068 | 124 | 124 | 4516 |
|  create.SimpleMutation | 7 | 7 | 48 | 63 | 1 |
|  cypher-directive.TopLevelMutationDirective | 1135 | 1135 | 46 | 35 | 1134 |
|  delete.SimpleDelete | 19401 | 19401 | 238 | 208 | 1040 |
|  delete.NestedDeleteInUpdate | 19165 | 19165 | 172 | 162 | 2040 |
|  2871.NestedRelationshipFilter | 23097 | 23086 | 55 | 70 | 4439 |
|  2925.SingleRelationshipFilter | 6468 | 6468 | 37 | 43 | 1040 |
|  2925.NestedSingleRelationshipFilter | 22891 | 22891 | 84 | 108 | 2174 |
|  2925.SingleRelationshipRequiredFilter | 5201 | 5201 | 37 | 43 | 1040 |
|  2925.NestedSingleRelationshipRequiredFilter | 9361 | 9361 | 50 | 84 | 1040 |
|  query.SimpleQuery | 3121 | 3121 | 19 | 32 | 1040 |
|  query.SimpleQueryWithRelationship | 16129 | 16129 | 39 | 54 | 2174 |
|  query.QueryWhere | 9684 | 9679 | 33 | 47 | 2164 |
|  query.SimpleQueryWithNestedWhere | 9851 | 9846 | 45 | 55 | 2164 |
|  query.Nested | 10095502 | 10095502 | 7129 | 6770 | 2008534 |
|  query.NestedWithFilter | 10074351 | 10074351 | 6939 | 6810 | 2004000 |
|  query.OrFilterOnRelationships | 42383 | 42064 | 143 | 173 | 1663 |
|  query.OrFilterOnRelationshipsAndNested | 36067 | 36394 | 191 | 242 | 1663 |
|  query.QueryWithNestedIn | 13501 | 13255 | 50 | 73 | 1262 |
|  query.NestedConnectionWhere | 9801 | 9801 | 57 | 67 | 2174 |
|  query.DeeplyNestedConnectionWhere | 9869 | 9804 | 66 | 85 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 18800 | 18686 | 113 | 158 | 1598 |
|  query.NestedWithRelationshipSingleFilters | 3865 | 3865 | 138 | 201 | 1134 |
|  query.Fulltext | 64 | 64 | 31 | 37 | 16 |
|  query.FulltextWithNestedQuery | 558 | 558 | 40 | 45 | 84 |
|  sorting-and-cypher.TopLevelSortWithCypher | 14059 | 14059 | 38 | 45 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithCypher | 14059 | 14059 | 57 | 68 | 2174 |
|  sorting-and-cypher.TopLevelSortWithCypherWithNested | 14204 | 14204 | 51 | 56 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithCypherWithNested | 14204 | 14204 | 91 | 98 | 2174 |
|  sorting-and-cypher.TopLevelSortWithExpensiveCypher | 14860 | 14924 | 91 | 114 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithExpensiveCypher | 14404 | 14404 | 101 | 128 | 2174 |
|  sorting.SortMultipleTypes | 3498 | 3498 | 82 | 92 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 14504 | 14504 | 92 | 119 | 2174 |
|  sorting.SortOnNestedFields | 14049 | 14049 | 43 | 74 | 2174 |
|  sorting.SortDeeplyNestedFields | 43148 | 43148 | 74 | 91 | 4516 |
|  sorting.ConnectionWithSort | 3281 | 3281 | 67 | 80 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 55 | 75 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 51 | 65 | 35 |
|  unions.NestedUnion | 410637 | 410637 | 231 | 262 | 33033 |
|  unions.NestedUnionWithMissingFields | 384611 | 384611 | 228 | 239 | 33033 |
|  update.NestedUpdate | 16143 | 16143 | 111 | 112 | 2002 |

</details>
        
Old Schema Generation: 26.506s
Schema Generation: 27.048s
Old Subgraph Schema Generation: 31.230s
Subgraph Schema Generation: 30.799s

<!-- Sticky Pull Request Commentperformance -->

mjfwebb committed:
docs: add changeset

mjfwebb committed:
Merge branch 'dev' into single-vector-directive

mjfwebb committed:
feat: rename propertyName to embeddingProperty

darrellwarde commented on .changeset/thirty-nails-shake.md:
```diff
@@ -0,0 +1,5 @@
+---
```
I think we can just have the "main" changelog entry for this?

mjfwebb replied:
@angrykoala thoughts?

angrykoala replied:
Fine by me

darrellwarde commented on packages/graphql/src/graphql/directives/vector.ts:
```diff
@@ -0,0 +1,76 @@
+    GraphQLString,
+} from "graphql";
+
+export const VectorProviderNames = ["VertexAI", "OpenAI", "AzureOpenAI", "Bedrock"] as const;
```
May be unpopular, but... This is a GraphQL enum - should these be capitalized as is conventional?

```suggestion
export const VectorProviderNames = ["VERTEX_AI", "OPEN_AI", "AZURE_OPEN_AI", "BEDROCK"] as const;
```

mjfwebb replied:
I would personally very much prefer to keep them as they are since they mirror the existing documentation and the configuration in the constructor

angrykoala replied:
I'm torn on this one, I'd say that we should match the enum standard in the rest of the API, but it is true that the constructor configuration should match this if possible :thinking:

darrellwarde commented on packages/graphql/src/schema/resolvers/query/vector.ts:
```diff
@@ -0,0 +1,102 @@
+        };
+
+        // (context as Neo4jGraphQLTranslationContext).resolveTree = resolveTree;
+
```
```suggestion
```

darrellwarde commented on packages/graphql/src/schema/resolvers/query/vector.ts:
```diff
@@ -0,0 +1,102 @@
+
+        return {
+            [entityAdapter.operations.rootTypeFieldNames.connection]: {
+                totalCount,
+                edges,
+                pageInfo,
```
Is it maybe better to have a constant somewhere which is like an "empty connection"? Then could just return that rather than defining with default values on like 69, and you could just return the connection as it is if there is a record in the result.

darrellwarde commented on packages/graphql/src/translate/queryAST/ast/selection/VectorSelection.ts:
```diff
@@ -0,0 +1,121 @@
+        // 1. Vector index without phrase, where the input is a List of Floats
+        if (this.vectorOptions.vector) {
+            vectorClause = Cypher.db.index.vector
+                .queryNodes(indexName, 10, vectorParam) // TODO: Check if 10 is a good default value
```
Have you checked this? 😅

darrellwarde commented on packages/graphql/src/translate/queryAST/factory/Operations/AggregateFactory.ts:
```diff
@@ -136,7 +136,10 @@ export class AggregateFactory {
             if (isConcreteEntity(entity)) {
                 let selection: EntitySelection;
                 if (context.resolveTree.args.fulltext || context.resolveTree.args.phrase) {
+                    // QUESTION: Why can't we just check for context.resolveTree.args.fulltext?
```
Is this a question that still needs answering?

darrellwarde commented on packages/graphql/src/translate/queryAST/factory/Operations/FulltextFactory.ts:
```diff
@@ -159,6 +162,8 @@ export class FulltextFactory {
             };
         }
 
+        // QUESTION: Is this ever reached?
```
Did you try deleting it to see if anything breaks? 😅

darrellwarde commented on packages/graphql/src/translate/queryAST/factory/Operations/VectorFactory.ts:
```diff
@@ -0,0 +1,206 @@
+            score: context.vector.scoreVariable,
+        };
+
+        // const entries = Object.entries(context.resolveTree.args.vector || {});
+        // if (entries.length > 1) {
+        //     throw new Error("Can only call one search at any given time");
+        // }
+        // const [indexName, indexInput] = entries[0] as [string, { phrase: string }];
+        // return {
+        //     index: indexName,
+        //     phrase: indexInput.phrase,
+        //     score: new Cypher.Variable(),
+        // };
+    }
```
?

```suggestion
```

darrellwarde reviewed (commented):
Looks good! A few comments to get the ball rolling!

mjfwebb committed:
Apply suggestions from code review

Co-authored-by: Darrell Warde <8117355+darrellwarde@users.noreply.github.com>

mjfwebb committed:
fix: re-add closing brace

mjfwebb committed:
docs: update comments

mjfwebb committed:
refactor: early return empty connection if there are no records

mjfwebb committed:
docs: remove additional changeset

mjfwebb committed:
docs: add license header

angrykoala commented on packages/graphql/src/classes/utils/asserts-indexes-and-constraints.ts:
```diff
@@ -82,12 +92,47 @@ async function getExistingIndexes({
     return existingIndexes;
 }
 
+function checkVectorIndexes(entity: ConcreteEntity, existingIndexes: ExistingIndexes, indexErrors: string[]) {
```
**Nitpick** Add `:void` to make it clear that this function does not return anything

angrykoala commented on packages/graphql/src/schema/resolvers/query/shared.ts:
```diff
@@ -0,0 +1,29 @@
+ * limitations under the License.
+ */
+
+export const emptyConnection = {
```
Why `shared.ts`? It is a bit confusing.

I think it makes a bit more sense to have the name of the file reflect what's in there (e.g. `empty-connection.ts`) if we want to make clear the intent of being a shared file maybe move it to a folder `common` or `utils` within the query folder?

mjfwebb replied:
Good point! `shared.ts` simply as it was shared between the vector resolver and root-connection resolver. I'll change it to `empty-connection.ts`.

angrykoala commented on packages/graphql/tests/tck/directives/vector/phrase.test.ts:
```diff
@@ -0,0 +1,109 @@
+describe("phrase input - genAI plugin", () => {
+    let typeDefs: string;
+    let neoSchema: Neo4jGraphQL;
+    let verifyTCK;
```
```suggestion
    let verifyTCK: boolean;
```

mjfwebb replied:
This isn't directly possible unfortunately as the type of `process.env.VERIFY_TCK` is a string. Also this would probably need to be changed holistically I think?

Perhaps we can look at this after this PR?

angrykoala replied:
that means that `VERIFY_TCK=false` would be treated as true which is not ideal.

Yeah, probably requires checking it everywhere

mjfwebb replied:
I'll follow this up

angrykoala commented on packages/graphql/tests/tck/directives/vector/score.test.ts:
```diff
@@ -0,0 +1,577 @@
+            typeDefs,
+        });
+
+        if (process.env.VERIFY_TCK) {
+            verifyTCK = process.env.VERIFY_TCK;
+            delete process.env.VERIFY_TCK;
+        }
```
Worth adding a comment as to why this test skips the tck verification?

angrykoala commented on packages/graphql/src/translate/queryAST/factory/FilterFactory.ts:
```diff
@@ -470,6 +486,26 @@ export class FilterFactory {
         });
     }
 
+    // private createScorePropertyFilter(
+    //     entity: ConcreteEntityAdapter,
+    //     isNot: boolean,
+    //     operator: WhereOperator | undefined,
+    //     value: string
+    // ): Filter {
+    //     // Value will be in shape { min: number, max: number } like { min: 0.5, max: 1.0 }
+    //     const scoreAttribute = entity.findAttribute("score");
+    //     if (!scoreAttribute) {
+    //         throw new Error(`score attribute not found`);
+    //     }
+
+    //     return this.createPropertyFilter({
+    //         attribute: scoreAttribute,
+    //         comparisonValue: "score",
+    //         isNot,
+    //         operator,
+    //     });
+    // }
+
```
Is this comment useful?

mjfwebb replied:
Not at all! :+1:

angrykoala reviewed (commented):
Looks good, a few nitpicks and minor comments

mjfwebb committed:
feat: add return types to assert index and constraints functions

mjfwebb committed:
refactor: rename shared to empty-connection

mjfwebb committed:
refactor: remove commented out code

mjfwebb committed:
docs: add note explaining why tck verification is skipped for vector

tck verification is skipped for vector tests as vector is not supported
on Neo4j 4.x which is the version of the DB we use when verifying tck

mjfwebb committed:
Merge branch 'dev' into single-vector-directive

mjfwebb committed:
fix: remove import file extension

mjfwebb committed:
refactor: use UPPER_CASE for the GraphQL provider name enums

mjfwebb committed:
feat: set 4 as the default value for nearestNeighbours

mjfwebb committed:
fix: update to use UPPER_CASE for OpenAI enum

mjfwebb committed:
fix: update snapshots to use 4 as the value for nearestNeighbours

mjfwebb committed:
Merge branch 'dev' into single-vector-directive
